DigitalLearningSolutions.Data/Services/RecommendedLearningService.cs

```diff
@@ -67,40 +67,111 @@ int delegateId
+            List<CompetencyLearningResource> competencyLearningResources
         )
         {
+            var learningLogItemsForResource = delegateLearningLogItems.Where(
+                ll => ll.ArchivedDate == null && ll.LearningHubResourceReferenceId == rr.RefId
+            ).ToList();
+            var incompleteLearningLogItem =
+                learningLogItemsForResource.SingleOrDefault(ll => ll.CompletedDate == null);
+
             var clrsForResource =
-                competencyLearningResources.Where(clr => clr.LearningHubResourceReferenceId == resource.RefId).ToList();
+                competencyLearningResources.Where(clr => clr.LearningHubResourceReferenceId == rr.RefId)
+                    .ToList();
 
             var competencyResourceAssessmentQuestionParameters =
                 competencyLearningResourcesDataService
-                    .GetCompetencyResourceAssessmentQuestionParameters(clrsForResource.Select(clr => clr.Id)).ToList();
+                    .GetCompetencyResourceAssessmentQuestionParameters(clrsForResource.Select(clr => clr.Id))
+                    .ToList();
+
+            return AreDelegateAnswersWithinRangeToDisplayResource(
+                clrsForResource,
+                competencyResourceAssessmentQuestionParameters,
+                selfAssessmentId,
+                delegateId
+            )
+                ? new RecommendedResource(
+                    learningHubResourceReferenceId,
+                    rr,
+                    incompleteLearningLogItem,
+                    learningLogItemsForResource.Any(ll => ll.CompletedDate != null),
+                    CalculateRecommendedLearningScore(
+                        rr,
+                        clrsForResource,
+                        competencyResourceAssessmentQuestionParameters,
+                        selfAssessmentId,
+                        delegateId
+                    )
+                )
+                : null;
+        }
+
+        private bool AreDelegateAnswersWithinRangeToDisplayResource(
+            List<CompetencyLearningResource> clrsForResource,
+            List<CompetencyResourceAssessmentQuestionParameter> competencyResourceAssessmentQuestionParameters,
+            int selfAssessmentId,
+            int delegateId
+        )
+        {
+            foreach (var competencyLearningResource in clrsForResource)
+            {
+                var delegateResults = selfAssessmentDataService
+                    .GetSelfAssessmentResultsForDelegateSelfAssessmentCompetency(
+                        delegateId,
+                        selfAssessmentId,
+                        competencyLearningResource.CompetencyId
+                    ).ToList();
+
+                var competencyResourceAssessmentQuestionParameter =
+                    competencyResourceAssessmentQuestionParameters.Single(
+                        qp => qp.CompetencyLearningResourceId == competencyLearningResource.Id
+                    );

```

Comment:
The database enforces a one-to-one-or-zero relationship between CompetencyResourceAssessmentQuestionParameters and CompetencyLearningResource so the `.Single()` here should be `SingleOrDefault()`.
In the case of null being returned, then no record exists, and thus the linked competency does not have a range specified. As such, we can return true, as we have at least one linked competency with no range specified.

Also minor, but the record (if found) will technically contain multiple "parameters" so we might need a better name for the singular record variable. I can't think of anything better than maybe "CompetencyParametersForAssessmentQuestion", so feel free to improvise.

Reply:
Changed the Single and added the null behaviour.
I'm not sure I fully agree on the name change but have changed it to competencyResourceAssessmentQuestionParametersForClr.
Plural nouns for a single table record isn't really great as it leaves you unable to make it singular to signify a single record